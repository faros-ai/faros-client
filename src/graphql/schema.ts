import {ok} from 'assert';
import _, {camelCase} from 'lodash';

import {
  ArrayForeignKey,
  ArrayRelationship,
  ManualConfiguration,
  ObjectForeignKey,
  ObjectRelationship,
  Schema,
} from './types';

export const MULTI_TENANT_COLUMNS = new Set(['tenant_id', 'graph_name']);
const PG_TYPE_REGEX = /\((.*)\)::.*/;

export function foreignKeyForObj(rel: ObjectRelationship): string {
  if (isManualConfiguration(rel.using)) {
    // for object rel, column_mapping contains one entry
    // mapping FK on host table to PK of target table
    const fkToPkMap = rel.using.manual_configuration.column_mapping;
    const fks = Object.keys(fkToPkMap);
    ok(
      fks.length === 1,
      `expected one foreign key in ${JSON.stringify(fkToPkMap)}`
    );
    return fks[0];
  }
  const arrFK = rel.using as ObjectForeignKey;
  const fk = arrFK.foreign_key_constraint_on;
  ok(typeof fk === 'string', `expected object fk on ${JSON.stringify(arrFK)}`);
  return fk;
}

export function foreignKeyForArray(rel: ArrayRelationship): string {
  if (isManualConfiguration(rel.using)) {
    // for array rel, column_mapping contains one entry
    // mapping PK of target table to FK on host table
    const pkToFkMap = rel.using.manual_configuration.column_mapping;
    const pks = Object.keys(pkToFkMap);
    ok(
      pks.length === 1,
      `expected one foreign key in ${JSON.stringify(pkToFkMap)}`
    );
    return _.get(pkToFkMap, pks[0]);
  }
  const arrFK = rel.using as ArrayForeignKey;
  const fk = arrFK.foreign_key_constraint_on?.column;
  ok(fk, `expected array fk on ${JSON.stringify(arrFK)}`);
  return fk;
}

export function foreignKeyForReverseObj(
  rel: ObjectRelationship
): any | undefined {
  if (isManualConfiguration(rel.using)) {
    // for reverse object rel, column_mapping contains one entry
    // mapping literal id to FK on remote table
    return {
      fkCol: rel.using.manual_configuration.column_mapping?.id,
      sourceTable: rel.using.manual_configuration.remote_table?.name
    };
  }
  return undefined;
}

/**
 * Parse elements of primary key from pkey function definition.
 * e.g. pkey(VARIADIC ARRAY[tenant_id, graph_name, source, uid])
 */
export function parsePrimaryKeys(
  exp: string,
  camelCaseFieldNames: boolean,
  includeMultiTenantColumns = false
): string [] {
  return exp
    .replace('pkey(VARIADIC ARRAY[', '')
    .replace('])', '')
    .split(', ')
    .map((col) => col.replace(/"/g, ''))
    .map((col) => {
      // extract col from types e.g. foo::text => foo
      const matches = col.match(PG_TYPE_REGEX);
      return matches ? matches[1] : col;
    })
    .filter((col) =>
      // conditionally filter multi-tenant columns
      includeMultiTenantColumns || !MULTI_TENANT_COLUMNS.has(col))
    .map((col) => (camelCaseFieldNames ? camelCase(col) : col));
}

export function isManualConfiguration(
  using: ManualConfiguration | any
): using is ManualConfiguration {
  return (using as ManualConfiguration).manual_configuration !== undefined;
}

export interface SchemaLoader {
  loadSchema(): Promise<Schema>;
}
