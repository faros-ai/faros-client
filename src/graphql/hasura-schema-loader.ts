import {ok} from 'assert';
import axios, {AxiosInstance} from 'axios';
import fs from 'fs-extra';
import {camelCase, find, isEmpty, snakeCase} from 'lodash';
import path from 'path';
import pino, {Logger} from 'pino';
import toposort from 'toposort';
import {assert, Dictionary} from 'ts-essentials';
import {VError} from 'verror';

import {
  foreignKeyForArray,
  foreignKeyForObj,
  foreignKeyForReverseObj,
  isManualConfiguration,
  MULTI_TENANT_COLUMNS,
  parsePrimaryKeys,
  SchemaLoader,
} from './schema';
import {
  ArrayForeignKey,
  ArrayRelationship,
  BackReference,
  ObjectRelationship,
  Reference,
  Schema,
  Table,
} from './types';

// Loads schema from a Hasura api
export class HasuraSchemaLoader implements SchemaLoader {
  private readonly api: AxiosInstance;

  /**
   * Create a schema loader
   * @param url Hasura instance URL
   * @param adminSecret Hasura admin credential
   * @param camelCaseFieldNames where applicable ensure names that will be
   * interpreted as GraphQL fields are camel cased. False by default
   * @param logger
   */
  constructor(
    url: string,
    adminSecret?: string,
    private readonly camelCaseFieldNames = false,
    private readonly logger: Logger = pino({name: 'hasura-schema-loader'}),
  ) {
    this.api = axios.create({
      baseURL: url,
      headers: {
        'X-Hasura-Role': 'admin',
        ...(adminSecret && {'X-Hasura-Admin-Secret': adminSecret}),
      },
    });
  }

  private async fetchDbSource(): Promise<Source> {
    const response = await this.api.post('/v1/metadata', {
      type: 'export_metadata',
      version: 2,
      args: {},
    });
    const sources: Source[] = response.data.metadata.sources;
    const defaultSource = find(sources, (source) => source.name === 'default');
    if (!defaultSource) {
      throw new VError('Faros database not connected to Hasura');
    }
    return defaultSource;
  }

  private async fetchPrimaryKeys(): Promise<Dictionary<string[]>> {
    const response = await this.api.post('/v2/query', {
      type: 'run_sql',
      args: {
        source: 'default',
        sql: await fs.readFile(
          path.join(__dirname, '../../resources/fetch-primary-keys.sql'),
          'utf8'
        ),
        cascade: false,
        read_only: true,
      },
    });
    const result: [string, string][] = response.data.result;
    if (!result) {
      throw new VError('Failed to load primary keys');
    }
    const primaryKeys: Dictionary<string[]> = {};
    result
      .filter((row) => row[0] !== 'table_name')
      .forEach(([table, exp]) => {
        primaryKeys[table] = parsePrimaryKeys(exp, this.camelCaseFieldNames);
      });
    return primaryKeys;
  }

  /**
   * Builds map from sourceTable to map of targetTable by fk column.
   *
   * That is, given:
   *   sourceTable (e.g. cicd_Build)
   *   fk column (e.g. pipeline_id in cicd_Build)
   *   targetTable (e.g. cicd_Pipeline): table.table.name
   *
   * The output, res, can be used as:
   *   res['cicd_Build']['pipeline_id'] => 'cicd_Pipeline'
   */
  static indexFkTargetModels(source: Source): Dictionary<Dictionary<string>> {
    const res: Dictionary<Dictionary<string>> = {};
    for (const table of source.tables) {
      const targetTable: string = table.table.name;
      for (const arrRel of table.array_relationships || []) {
        const fkCol = foreignKeyForArray(arrRel);
        const sourceTable = remoteTableForArray(arrRel);
        ok(sourceTable, `missing source table on ${JSON.stringify(arrRel)}`);
        if (!res[sourceTable]) {
          res[sourceTable] = {};
        }
        res[sourceTable][fkCol] = targetTable;
      }
      // find array relationship represented as reverse object relationships
      for (const objRel of table.object_relationships || []) {
        const reverseObjFk = foreignKeyForReverseObj(objRel);
        if (reverseObjFk) {
          if (!res[reverseObjFk.sourceTable]) {
            res[reverseObjFk.sourceTable] = {};
          }
          res[reverseObjFk.sourceTable][reverseObjFk.fkCol] = targetTable;
        }
      }
    }
    return res;
  }

  async loadSchema(): Promise<Schema> {
    const primaryKeysFromDb = await this.fetchPrimaryKeys();
    const source = await this.fetchDbSource();
    const targetTableByFk = HasuraSchemaLoader.indexFkTargetModels(source);
    const query = await fs.readFile(
      path.join(__dirname, '../../resources/introspection-query.gql'),
      'utf8'
    );
    const response = await this.api.post('/v1/graphql', {query});
    const schema = response.data.data.__schema;
    const tableNames = [];
    const scalars: Dictionary<Dictionary<string>> = {};
    const references: Dictionary<Dictionary<Reference>> = {};
    const backReferences: Dictionary<BackReference[]> = {};
    const primaryKeysFromSchema: Dictionary<string []> = {};
    for (const table of source.tables) {
      const tableName = table.table.name;
      tableNames.push(tableName);
      const type = find(
        schema.types,
        (t) => t.name === tableName && t.kind === 'OBJECT'
      );
      if (!type?.fields) {
        continue;
      }
      const scalarTypes: any[] = type.fields.filter(
        (t: any) => this.unwrapType(t.type).kind === 'SCALAR'
          && t.description !== 'generated'
      );
      const tableScalars: Dictionary<string> = {};
      for (const scalar of scalarTypes) {
        if (!MULTI_TENANT_COLUMNS.has(snakeCase(scalar.name))) {
          tableScalars[scalar.name] = this.unwrapType(scalar.type).name;
        }
      }
      scalars[tableName] = tableScalars;
      if (type.description) {
        try {
          const {primaryKeys} = JSON.parse(type.description);
          if (primaryKeys) {
            assert(Array.isArray(primaryKeys), 'primaryKeys is not an array');
            primaryKeysFromSchema[tableName] = this.camelCaseFieldNames ?
              primaryKeys.map((c) => camelCase(c)) : primaryKeys;
          }
        } catch (e) {
          this.logger.debug(
            `unable to parse ${tableName} description: ${JSON.stringify(e)}`
          );
        }
      }
      const tableReferences: Dictionary<Reference> = {};
      for (const rel of table.object_relationships ?? []) {
        const fk = foreignKeyForObj(rel);
        // skip reverse object relationships
        if (fk === 'id') {
          continue;
        }
        const relFldName = this.camelCaseFieldNames ? camelCase(fk) : fk;
        const relMetadata = {
          field: rel.name,
          model: targetTableByFk[table.table.name][fk],
          foreignKey: relFldName
        };
        // index relation metadata using both FK column and rel.name
        // this is needed for cross-compatibility with CE and SaaS
        // GraphQLClient iterates over fields of records checking this
        // data structure for matches on nested objects.
        // the record contains SDL relationship names like owner or repository
        // in CE, the FK column matches this relationship name
        // in SaaS, the column is irrelevant we need to index by Hasura rel.name
        tableReferences[relFldName] = relMetadata;
        tableReferences[rel.name] = relMetadata;
      }
      references[tableName] = tableReferences;
      backReferences[tableName] = (table.array_relationships ?? []).map(
        (rel) => {
          return {
            field: rel.name,
            model: remoteTableForArray(rel),
          };
        }
      );
      const reverseBackRefs: BackReference[] = [];
      (table.object_relationships ?? []).forEach(
        (rel) => {
          const reverseObjFk = foreignKeyForReverseObj(rel);
          if (reverseObjFk) {
            reverseBackRefs.push({
              field: rel.name,
              model: reverseObjFk.sourceTable,
            });
          }
        },
      );
      backReferences[tableName].push(...reverseBackRefs);
    }
    const modelDeps: [string, string][] = [];
    for (const model of Object.keys(references)) {
      for (const ref of Object.values(references[model])) {
        if (model !== ref.model) {
          modelDeps.push([model, ref.model]);
        }
      }
    }
    const sortedModelDependencies = toposort(modelDeps);
    return {
      primaryKeys: isEmpty(primaryKeysFromSchema) ?
        primaryKeysFromDb : primaryKeysFromSchema,
      scalars,
      references,
      backReferences,
      sortedModelDependencies,
      tableNames,
    };
  }

  // Unwraps type from its wrapping type (non-null and list containers)
  private unwrapType(type: any): any {
    if (type.kind === 'LIST' || type.kind === 'NON_NULL') {
      return this.unwrapType(type.ofType);
    }
    return type;
  }
}

function remoteTableForArray(rel: ArrayRelationship): string {
  if (isManualConfiguration(rel.using)) {
    return rel.using.manual_configuration.remote_table.name;
  }
  const arrFK = rel.using as ArrayForeignKey;
  const table = arrFK.foreign_key_constraint_on?.table?.name;
  ok(table, `expected remote table on ${JSON.stringify(arrFK)}`);
  return table;
}

interface TableWithRelationships {
  table: Table;
  object_relationships: ReadonlyArray<ObjectRelationship>;
  array_relationships: ReadonlyArray<ArrayRelationship>;
}

interface Source {
  name: string;
  kind: string;
  tables: ReadonlyArray<TableWithRelationships>;
  configuration: any;
}
