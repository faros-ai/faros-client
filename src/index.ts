export {
  DEFAULT_RETRY_DELAY,
  DEFAULT_RETRIES,
  makeAxiosInstance,
  makeAxiosInstanceWithRetry,
} from './axios';
export {wrapApiError} from './errors';
export {FarosClient} from './client';
export {
  Account,
  Address,
  Coordinates,
  FarosClientConfig,
  Location,
  Model,
  NamedQuery,
  SecretName,
  UpdateAccount,
} from './types';
export {
  foreignKeyForArray,
  foreignKeyForObj,
  SchemaLoader,
} from './graphql/schema';
export {
  ArrayForeignKey,
  ArrayRelationship,
  ManualConfiguration,
  ObjectForeignKey,
  ObjectRelationship,
  Reference,
  BackReference,
  Schema,
  PathToModel,
  Query,
} from './graphql/types';
export {HasuraSchemaLoader} from './graphql/hasura-schema-loader';
export {
  AnyRecord,
  FlattenContext,
  PaginatedQuery,
  Reader,
  RecordIterable,
  buildIncrementalQueryV1,
  buildIncrementalQueryV2,
  createIncrementalQueriesV1,
  createIncrementalQueriesV2,
  createIncrementalReadersV1,
  createIncrementalReadersV2,
  createNonIncrementalReaders,
  crossMerge,
  flatten,
  flattenIterable,
  flattenV2,
  paginatedQuery,
  paginatedQueryV2,
  pathToModelV1,
  pathToModelV2,
  queryNodesPaths,
  readerFromQuery,
  toIncrementalV1,
  toIncrementalV2,
} from './graphql/graphql';
export {Utils} from './utils';
