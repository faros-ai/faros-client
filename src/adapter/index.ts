import * as gql from 'graphql';
import {Maybe} from 'graphql/jsutils/Maybe';
import _ from 'lodash';
import {DateTime} from 'luxon';
import {singular} from 'pluralize';
import {Writable} from 'ts-essentials';
import VError from 'verror';

import {FarosClient} from '../client';
import {
  isEmbeddedObjectListType,
  isEmbeddedObjectType,
  isModelQuery,
  isObjectListType,
  METADATA,
  NODES,
  paginatedQueryV2,
  REFRESHED_AT
} from '../graphql/graphql';
import {
  asLeafValueType,
  FieldPaths,
  isLeafType,
  isLeafValue,
  isNestedValue,
  isPrimitiveListType,
  isTypedArray,
  LeafValueType,
} from './types';

export {FieldPaths} from './types';

/**
 * Converts a V1 query namespace, specified as a model namespace and model name,
 * into a V2 query namespace.
 *
 * For example, queryNamespace('vcs', 'commits') returns 'vcs_Commit'.
 */
function queryNamespace(modelNamespace: string, model: string): string {
  let modelName;
  if (modelNamespace === 'faros' && model.endsWith('Options')) {
    // Don't singularize options models
    modelName = model;
  } else {
    modelName = singular(model);
  }
  return `${modelNamespace}_${_.upperFirst(modelName)}`;
}

/**
 * Returns the V2 field for an embedded V1 field.
 *
 * For example, embeddedName('env', 'category') returns 'envCategory'.
 */
function embeddedName(parentName: string, fieldName: string): string {
  return `${parentName}${_.upperFirst(fieldName)}`;
}

/** Should this V1 field be omitted from a V2 query? */
function omitField(name: string, type: gql.GraphQLOutputType): boolean {
  return name === METADATA || name === NODES || isEmbeddedObjectType(type);
}

/** Extracts nested fields with optional field rename */
function nestedFields(
  field: gql.FieldNode,
  rename?: (name: string) => string
): gql.FieldNode[] {
  const nestedFields: gql.FieldNode[] = [];
  const selections = field.selectionSet?.selections ?? [];
  for (const selection of selections) {
    if (selection.kind === 'Field') {
      const name = selection.name.value;
      nestedFields.push({
        ...selection,
        name: {
          kind: gql.Kind.NAME,
          value: rename ? rename(name) : name
        }
      });
    }
  }
  return nestedFields;
}

/**
 * Returns a mapping from V2 to V1 field paths. If a field is an object list,
 * then its path will point to a nested structure that contains paths that
 * are relative to it.
 *
 * Note: In V2, fields inside embedded object lists are stored identically to
 * those in V1. The "mirrorPaths" boolean argument is true whenever traversal is
 * happening inside an emdedded object list. This causes the V1 and V2 paths to
 * be set to the same values, i.e., they'll be "mirrored".
 */
export function getFieldPaths(
  ast: gql.ASTNode,
  typeInfo: gql.TypeInfo,
  mirrorPaths = false
): FieldPaths {
  const fieldStack: string[] = [];
  const newFieldStack: string[] = [];
  function pushField(name: string, type: gql.GraphQLOutputType): void {
    fieldStack.push(name);
    if (mirrorPaths || !omitField(name, type)) {
      newFieldStack.push(name);
    }
  }
  function popField(name: string): void {
    fieldStack.pop();
    if (mirrorPaths || _.last(newFieldStack) === name) {
      newFieldStack.pop();
    }
  }

  const fieldPaths: Writable<FieldPaths> = {};
  gql.visit(ast, gql.visitWithTypeInfo(typeInfo, {
    Field: {
      enter(node) {
        let type = typeInfo.getType();
        const parentType = typeInfo.getParentType();
        if (gql.isNonNullType(type)) {
          type = type.ofType;
        }
        if (!type) {
          throw new VError(
            'unable to determine type of field: %s',
            fieldStack.join('.')
          );
        }

        const fieldName = node.name.value;
        pushField(fieldName, type);
        if (isModelQuery(parentType, type)) {
          // Convert the V1 query namespace to a V2 query namespace
          // For example: vcs { commits { ...} } => vcs_Commit { ... }
          newFieldStack.shift();
          newFieldStack.shift();
          const [namespace, name] = fieldStack;
          newFieldStack.unshift(queryNamespace(namespace, name));
        } else if (isObjectListType(type)) {
          // Recursive call to next object list
          const nextNode = node.selectionSet;
          if (nextNode) {
            const newFieldPath = newFieldStack.join('.');
            const fieldPath = fieldStack.join('.');
            typeInfo.enter(nextNode);
            fieldPaths[newFieldPath] = {
              path: fieldPath,
              nestedPaths: getFieldPaths(
                nextNode,
                typeInfo,
                mirrorPaths || isEmbeddedObjectType(type.ofType)
              )
            };
            typeInfo.leave(nextNode);
            popField(fieldName);
            return false;
          }
        } else if (isLeafType(type)) {
          const fieldPath = fieldStack.join('.');
          let newFieldPath = newFieldStack.join('.');
          // Timestamps in v1 are always stringified epoch millis, except when
          // they're inside embedded object lists. In that case, they're stored
          // as epoch millis.
          const stringifyTimestamps = !mirrorPaths;
          let fieldType = asLeafValueType(type, stringifyTimestamps);
          if (mirrorPaths) {
            fieldPaths[fieldPath] = {path: fieldPath, type: fieldType};
            return undefined;
          } else if (isEmbeddedObjectType(parentType)) {
            const parentName = fieldStack[fieldStack.length - 2];
            if (parentName === METADATA) {
              if (fieldName === REFRESHED_AT) {
                // Hack: While V1 serializes "refreshedAt" the same as other
                // timestamps (epoch millis string), it types it as a string.
                // In V2, it's stored and typed like every other timestamp.
                // We force conversion from ISO 8601 string => epoch millis
                // string by overriding the type from string to timestamp.
                fieldType = 'epoch_millis_string';
              }
            } else {
              // Prefix the last field name with the embedded field name
              newFieldPath = [
                ...newFieldStack.slice(0, -1),
                embeddedName(parentName, fieldName)
              ].join('.');
            }
          }
          fieldPaths[newFieldPath] = {path: fieldPath, type: fieldType};
        }
        return undefined;
      },
      leave(node) {
        popField(node.name.value);
      }
    }
  }));
  return fieldPaths;
}

/**
 * Converts a V1 query into a V2 query using the following rules:
 *
 * 1. Replaces the first two field selections, i.e., the model namespace and
 *    name, are replaced by a single, combined selection. Example:
 *
 *    vcs { commits { sha createdAt } } => vcs_Commit { sha createdAt }
 *
 * 2. Flattens the selection set of each "nodes" and "metadata" field into the
 *    selection set of its parent. Examples:
 *
 *    commits { nodes { sha createdAt } } => commits { sha createdAt }
 *    commit { metadata { refreshedAt } } => commit { refreshedAt }
 *
 * 3. Flattens the selection set of each embedded field into the selection set
 *    of its parent, prefixing names of nested fields with the parent name.
 *    Example:
 *
 *    deployment { env { category } } => deployment { envCategory }
 *
 * 4. Removes nested fields from embedded object lists since these are stored
 *    as JSONB arrays in V2 and cannot be extracted. Example:
 *
 *    task { additionalFields { name value } } => task { additionalFields }
 *
 * 5. Renames the "first" field argument to "limit". Example:
 *
 *    deployments(first: 1) { uid } => deployments(limit: 1) { uid }
 */
export function asV2AST(ast: gql.ASTNode, typeInfo: gql.TypeInfo): gql.ASTNode {
  return gql.visit(ast, gql.visitWithTypeInfo(typeInfo, {
    Field: {
      // Handles rule (1)
      leave(node, key, parent, path, ancestors) {
        const grandparent = ancestors[ancestors.length - 2];
        if (isTypedArray<gql.ASTNode>(grandparent)) {
          return undefined;
        } else if (grandparent.kind !== 'OperationDefinition') {
          return undefined;
        } else if (grandparent.operation !== 'query') {
          throw new Error('only queries can be converted');
        }

        const modelField = node.selectionSet?.selections?.[0];
        if (!modelField || modelField.kind !== 'Field') {
          throw new Error('query does not select a model');
        }
        const modelNamespace = node.name.value;
        const model = modelField.name.value;
        return {
          ...modelField,
          name: {
            kind: gql.Kind.NAME,
            value: queryNamespace(modelNamespace, model)
          },
        };
      }
    },
    // Handles rules (2), (3) and (4)
    SelectionSet: {
      leave(node) {
        const newSelections: gql.SelectionNode[] = [];
        for (const selection of node.selections) {
          typeInfo.enter(selection);
          const selectionType = typeInfo.getType();
          if (selection.kind === 'Field') {
            if (
              selection.name.value === NODES ||
              selection.name.value === METADATA
            ) {
              // Rule (2): flatten nodes and metadata fields
              newSelections.push(...nestedFields(selection));
            } else if (isEmbeddedObjectType(selectionType)) {
              // Rule (3): flatten embedded fields and rename them
              const prefix = selection.name.value;
              newSelections.push(...nestedFields(
                selection,
                (name) => embeddedName(prefix, name)
              ));
            } else if (isEmbeddedObjectListType(selectionType)) {
              // Rule (4): omit fields from embedded object lists
              newSelections.push(_.omit(selection, 'selectionSet'));
            } else {
              // Otherwise, leave the nested fields alone
              newSelections.push(selection);
            }
          }
          typeInfo.leave(selection);
        }
        return {kind: gql.Kind.SELECTION_SET, selections: newSelections};
      }
    },
    // Handles rule (5)
    Argument(node) {
      if (node.name.value === 'first') {
        return {...node, name: {kind: gql.Kind.NAME, value: 'limit'}};
      }
      return undefined;
    }
  }));
}

/** Shim that retrieves data from a V2 graph using a V1 query */
export class QueryAdapter {
  constructor(
    private readonly faros: FarosClient,
    private readonly v1Schema: gql.GraphQLSchema,
    private readonly v2Schema: Maybe<gql.GraphQLSchema> = undefined
  ) {
    if (faros.graphVersion !== 'v2') {
      throw new VError(
        'query adapter only works with v2 clients. found version: %s',
        faros.graphVersion
      );
    }
  }

  /** Converts a V2 field value into a V1 field value */
  private v1Value(v2Value: any, type: LeafValueType): any {
    if (isPrimitiveListType(type)) {
      if (Array.isArray(v2Value)) {
        // Recursively convert entries
        const v1List: any[] = [];
        for (const v2EntryValue of v2Value) {
          v1List.push(this.v1Value(v2EntryValue, type.entryType));
        }
        return v1List;
      }
      return v2Value;
    } else if (_.isNil(v2Value)) {
      return v2Value;
    }

    if (type === 'float' || type === 'double') {
      if (_.isString(v2Value)) {
        const double = parseFloat(v2Value);
        if (!isNaN(double)) {
          return double;
        }
      } else if (_.isNumber(v2Value)) {
        return v2Value;
      }
      throw new VError('invalid double: %s', v2Value);
    } else if (type === 'long') {
      // Long may be a string or number in V2
      if (_.isString(v2Value)) {
        if (/^-?\d+$/.test(v2Value)) {
          return v2Value;
        }
      } else if (_.isNumber(v2Value)) {
        return `${v2Value}`;
      }
      throw new VError('invalid long: %s', v2Value);
    } else if (type === 'epoch_millis' || type === 'epoch_millis_string') {
      const stringify = type === 'epoch_millis_string';
      if (_.isString(v2Value)) {
        const millis = DateTime.fromISO(v2Value).toMillis();
        if (!isNaN(millis)) {
          return stringify ? `${millis}` : millis;
        }
      } else if (_.isNumber(v2Value)) {
        return stringify ? `${v2Value}` : v2Value;
      }
      throw new VError('invalid timestamp: %s', v2Value);
    }
    return v2Value;
  }

  /** Converts a V2 node into a V1 node */
  private v1Node(v2Node: any, fieldPaths: FieldPaths): any {
    const v1Node: any = {};
    for (const [v2Path, v1Path] of Object.entries(fieldPaths)) {
      if (isLeafValue(v1Path)) {
        try {
          const v1Value = this.v1Value(_.get(v2Node, v2Path), v1Path.type);
          _.set(v1Node, v1Path.path, v1Value);
        } catch (err: any) {
          throw new VError(
            err,
            'failed to convert value in v2 field \'%s\' into value in v1 ' +
              'field \'%s\' of type \'%s\'',
            v2Path,
            v1Path.path,
            isPrimitiveListType(v1Path.type)
              ? `[${v1Path.type.entryType}]`
              : v1Path.type
          );
        }
      } else {
        const nestedV1Nodes: any[] = [];
        const nestedV2Nodes = _.get(v2Node, v2Path);
        const nestedPaths = v1Path.nestedPaths;
        if (Array.isArray(nestedV2Nodes)) {
          for (const nestedV2Node of nestedV2Nodes) {
            const nestedV1Node = this.v1Node(nestedV2Node, nestedPaths);
            if (nestedV1Node) {
              nestedV1Nodes.push(nestedV1Node);
            }
          }
        }
        _.set(v1Node, v1Path.path, nestedV1Nodes);
      }
    }
    return v1Node;
  }

  /** Returns paths relative to the initial nodes path */
  private nodePathsV1(
    v1AST: gql.ASTNode,
    v1TypeInfo: gql.TypeInfo
    ): FieldPaths {
    const fieldPaths = getFieldPaths(v1AST, v1TypeInfo);
    const [pathValue] = Object.values(fieldPaths);
    if (isNestedValue(pathValue)) {
      return pathValue.nestedPaths;
    }
    throw new VError('invalid path value: %s', pathValue);
  }

  private nodePathsV2(
    v2AST: gql.ASTNode,
    v2TypeInfo: gql.TypeInfo
  ): FieldPaths {
    const fieldPaths = getFieldPaths(v2AST, v2TypeInfo);
    console.log(fieldPaths);
    const [pathValue] = Object.values(fieldPaths);
    if (isNestedValue(pathValue)) {
      return pathValue.nestedPaths;
    }
    console.log(pathValue);
    throw new VError('invalid path value: %s', pathValue);
  }

  nodes(
    graph: string,
    query: string,
    pageSize = 100,
    args: Map<string, any> = new Map<string, any>(),
    postProcessV2Query: (v2Query: string) => string = _.identity
  ): AsyncIterable<any> {
        // Returns an object with a default async iterator
    // We try gql validation against v2 schema if v1 schema fails
    const queryAST = gql.parse(query);
    const v1ValidationErrors = gql.validate(this.v1Schema, queryAST);
    let v2Nodes: AsyncIterable<any>;
    if (this.v2Schema && v1ValidationErrors.length > 0) {
      const v2ValidationErrors = gql.validate(this.v2Schema, queryAST);
      if (v2ValidationErrors.length > 0) {
        throw new VError(
          'invalid query: %s\nValidation errors: %s',
          query,
          v1ValidationErrors.map((err) => err.message).join(', ') +
            v2ValidationErrors.map((err) => err.message).join(', ')
        );
      }
      v2Nodes = this.faros.nodeIterable(
        graph,
        query,
        pageSize,
        paginatedQueryV2,
        args
      );
      return {
        async *[Symbol.asyncIterator](): AsyncIterator<any> {
          for await (const v2Node of v2Nodes) {
            yield v2Node;
          }
        },
      };
    }

    const v1TypeInfo = new gql.TypeInfo(this.v1Schema);
    const v2Query = postProcessV2Query(
      gql.print(asV2AST(queryAST, v1TypeInfo))
    );
    const nodePaths = this.nodePathsV1(queryAST, v1TypeInfo);
    v2Nodes = this.faros.nodeIterable(
      graph,
      v2Query,
      pageSize,
      paginatedQueryV2,
      args
    );
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const self = this;
    return {
      async *[Symbol.asyncIterator](): AsyncIterator<any> {
        for await (const v2Node of v2Nodes) {
          yield self.v1Node(v2Node, nodePaths);
        }
      },
    };
  }
}
