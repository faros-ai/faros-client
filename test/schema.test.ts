import fs from 'fs';
import {buildSchema} from 'graphql';

import {FarosGraphSchema} from '../src/schema';

describe('schema', () => {
  const schema = new FarosGraphSchema(
    buildSchema(fs.readFileSync('test/resources/timestamp-schema.gql', 'utf8'))
  );

  test('ignore non-timestamp field', () => {
    const record = {uid: 'abc'};
    expect(schema.fixTimestampFields(record, 'ims_Incident')).toEqual(record);
  });

  test('fix timestamp field', () => {
    const record = {uid: 'abc', createdAt: new Date(123)};
    const json1 = {...record, createdAt: 123};
    const json2 = {...record, createdAt: '123'};
    const json3 = {...record, createdAt: record.createdAt.toISOString()};
    expect(schema.fixTimestampFields(json1, 'ims_Incident').createdAt).toEqual(
      new Date(123)
    );
    expect(schema.fixTimestampFields(json2, 'ims_Incident').createdAt).toEqual(
      new Date(123)
    );
    expect(schema.fixTimestampFields(json3, 'ims_Incident').createdAt).toEqual(
      new Date(123)
    );
  });

  test('fix non-null timestamp field', () => {
    const record = {uid: 'abc', createdAt: new Date(123)};
    const json = {...record, createdAt: 123};
    expect(schema.fixTimestampFields(json, 'ims_Incident').createdAt).toEqual(
      new Date(123)
    );
  });

  test('fix nested timestamp field', () => {
    const record = {uid: 'abc', timestampObj: {createdAt: new Date(123)}};
    const json = {uid: 'abc', timestampObj: {createdAt: 123}};
    expect(schema.fixTimestampFields(json, 'ims_Incident')).toEqual(record);
  });

  test('fix array of timestamp fields', () => {
    const record = {
      uid: 'abc',
      array: [
        {createdAt: new Date(123)},
        {createdAt: new Date(456)},
        {createdAt: new Date(789)},
      ],
    };
    const json = {
      uid: 'abc',
      array: [
        {createdAt: 123},
        {createdAt: '456'},
        {createdAt: new Date(789).toISOString()},
      ],
    };
    expect(schema.fixTimestampFields(json, 'ims_Incident')).toEqual(record);
  });

  test('fix timestamptz field', () => {
    const date = new Date(123);
    const record = {uid: 'abc', createdAt: date};
    const json1 = {...record, createdAt: 123};
    const json2 = {...record, createdAt: '123'};
    const json3 = {...record, createdAt: date.toISOString()};
    expect(schema.fixTimestampFields(json1, 'vcs_Commit').createdAt).toEqual(
      date.toISOString()
    );
    expect(schema.fixTimestampFields(json2, 'vcs_Commit').createdAt).toEqual(
      date.toISOString()
    );
    expect(schema.fixTimestampFields(json3, 'vcs_Commit').createdAt).toEqual(
      date.toISOString()
    );
  });

  test('invalid date in timestamptz field', () => {
    expect(() =>
      schema.fixTimestampFields({uid: 'abc', createdAt: 'bad'}, 'vcs_Commit')
    ).toThrowError('Invalid date: bad');

    expect(() =>
      schema.fixTimestampFields({uid: 'abc', createdAt: 10 ** 20}, 'vcs_Commit')
    ).toThrowError();
  });
});
