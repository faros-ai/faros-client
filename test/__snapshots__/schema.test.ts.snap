// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema fix jsonb - array 1`] = `
Object {
  "status": Array [
    Object {
      "changedAt": "2021-08-23T20:51:07.746Z",
      "foo": Object {
        "category": "Todo",
        "detail": "To Do",
        "fixAt": "2021-10-09T03:57:47.746Z",
        "skippedAt": "not-interesting",
      },
      "ignoredAt": "blah",
    },
    Object {
      "bar": Object {
        "category": "Todo",
        "detail": "Selected for Development",
        "fixAt": "2022-02-25T01:17:47.746Z",
        "skippedAt": "not-interesting",
      },
      "changedAt": "2021-08-24T19:16:00.991Z",
      "skipAt": "nothing",
    },
  ],
  "uid": "abc",
}
`;

exports[`schema fix jsonb - object 1`] = `
Object {
  "status": Object {
    "changedAt": "2021-08-23T20:51:07.746Z",
    "foo": Object {
      "ackedAt": "ignore",
      "brokenAt": "2022-02-13T11:31:07.746Z",
      "category": "Todo",
      "detail": "To Do",
    },
    "submittedAt": "foo",
  },
  "uid": "abc",
}
`;
